### PR TITLE
[GRIFFIN-358] Rewrite the Rule/ Measure implementations

### PR DESCRIPTION
**Pull request by chitralverma**
**What changes were proposed in this pull request?**

Current `RuleParams` can be of the following 3 DSL types,

- Data Ops (for source preprocessing)
- Griffin DSL
- SparkSQL

GriffinDSL allows the implementation of measures (DQ Types) like Completeness, Accuracy, etc.

To enable such measures there is an extensive implementation of expression, task hierarchies, parsing and most of this is heavily dependent on scala-parser-combinators.

At the end of the implementation, Griffin DSL tries to mimic a SparkSQL-like query but substitution of user-defined constraints.

This approach has some drawbacks,

- Suboptimal processing. While the transformation steps execute in parallel on the driver, the data set is still scanned multiple times in parallel which can cause inefficiencies on the SparkSession side and the internal task scheduler was single-threaded. Even though the data set can be cached, still it branched and crucial memory is required for holding the dataset rather than processing it.
- Internal functions of Spark are not used. Data preprocessing has a very limited scope currently even though we have 100s spark SQL functions available for use.
- This blocks structured streaming. The manually constructed SQL queries cause multiple aggregations in the same query on a streaming data set which is not supported by Spark's Structured streaming. There are workarounds for this but they all require rewriting the *Expr2DQSteps classes.
- Griffin DSL is SparkSQL like but not 100% compatible. Profiling measure and SparkSQL are redundant functionalities

The proposed solution involves SparkSQL DSL based measures and some changes to Rule Params. This will enhance the data pre proc flows and the measures themselves

**Does this PR introduce any user-facing change?**
Yes. Users can use the new measures as a separate configuration and there is scope for more data pre-processing.

**How was this patch tested?**
Unit Tests

### CODE REVIEW

**Comment by chitralverma** [comment]
@wankunde @guoyuepeng Can you please review this. thanks

File: measure/src/main/scala/org/apache/griffin/measure/datasource/connector/batch/ElasticSearchGriffinDataConnector.scala [hunk 1]
```diff
@@ -19,13 +19,13 @@ package org.apache.griffin.measure.datasource.connector.batch
...
 import java.io.{BufferedReader, ByteArrayInputStream, InputStreamReader}
 import java.net.URI
-import java.util
 
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
 import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
 import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import java.util
```
[the remark this answers is not in the document]
Comment (chitralverma) [reply measure/src/main/scala/org/apache/griffin/measure/datasource/connector/batch/ElasticSearchGriffinDataConnector.scala]: done.

**Comment by guoyuepeng** [comment]
big patch.
let me go through it today.

Thanks.

**Comment by chitralverma** [comment]
> big patch.
> let me go through it today.
> 
> Thanks.

Thanks.

**Comment by guoyuepeng** [comment]
LGTM.
Will merge it

**Review by guoyuepeng — APPROVED** [review]
LGTM

**Comment by chitralverma** [comment]
Thanks for the merge! :)

**Pull request merged**